airbyte-workers/src/main/java/io/airbyte/workers/DefaultReplicationWorker.java

```diff
@@ -89,41 +102,89 @@ public StandardSyncOutput run(StandardSyncInput syncInput, Path jobRoot) throws
       // closed first (which is what we want).
       try (destination; source) {
         destination.start(destinationConfig, jobRoot);
         source.start(sourceConfig, jobRoot);
 
-        while (!cancelled.get() && !source.isFinished()) {
-          final Optional<AirbyteMessage> maybeMessage = source.attemptRead();
-          if (maybeMessage.isPresent()) {
-            final AirbyteMessage message = mapper.mapMessage(maybeMessage.get());
-
-            messageTracker.accept(message);
-            destination.accept(message);
+        final Thread replicationThread = new Thread(() -> {
+          try {
+            while (!cancelled.get() && !source.isFinished()) {
+              final Optional<AirbyteMessage> messageOptional = source.attemptRead();
+              if (messageOptional.isPresent()) {
+                final AirbyteMessage message = mapper.mapMessage(messageOptional.get());
+
+                LOGGER.info("record in DefaultReplicationWorker: {}", message);
+
+                sourceMessageTracker.accept(message);
+                destination.accept(message);
+              }
+            }
+            destination.notifyEndOfStream();
+          } catch (Exception e) {
+            throw new RuntimeException(e);
           }
-        }
+        });
+
+        final Thread destinationOutputThread = new Thread(() -> {
+          try {
+            while (!cancelled.get() && !destination.isFinished()) {
+              final Optional<AirbyteMessage> messageOptional = destination.attemptRead();
+              if (messageOptional.isPresent()) {
+                LOGGER.info("state in DefaultReplicationWorker from Destination: {}", messageOptional.get());
+                destinationMessageTracker.accept(messageOptional.get());
+              }
+            }
+          } catch (Exception e) {
+            throw new RuntimeException(e);
+          }
+        });
+
+        final Future<?> destinationThreadExecution = executorService.submit(destinationOutputThread);
+        final Future<?> sourceThreadExecution = executorService.submit(replicationThread);
+
+        LOGGER.info("Waiting for source thread to join.");
+        sourceThreadExecution.get();
+        LOGGER.info("Source thread complete.");
+        LOGGER.info("Waiting for destination thread to join.");
+        destinationThreadExecution.get();
+        LOGGER.info("Destination thread complete.");
+
+      } catch (Exception e) {
+        hasFailed.set(true);
+        LOGGER.error("Sync worker failed.", e);
+      } finally {
+        executorService.shutdownNow();
       }
 
-      final StandardSyncSummary summary = new StandardSyncSummary()
-          .withStatus(cancelled.get() ? Status.FAILED : Status.COMPLETED)
-          .withRecordsSynced(messageTracker.getRecordCount())
-          .withBytesSynced(messageTracker.getBytesCount())
+      final ReplicationStatus outputStatus =
```

Comment:
NIT: might be a bit cleaner with an actual `if then if then else`

Reply:
I agree with using if then statements here.